tinygrad/runtime/support/tinymath.h

```diff
@@ -0,0 +1,58 @@
+#ifndef TINYMATH_H
+#define TINYMATH_H
+
+#include <stdint.h>
+
+#define NAN __builtin_nanf("")
+#define INFINITY __builtin_inff()
+
+static inline double sqrt(double x) {
+#if defined(__aarch64__) || defined(_M_ARM64)
+  __asm__("fsqrt %d0, %d1" : "=w"(x) : "w"(x));
+#elif defined(__x86_64__) || defined(_M_X64)
+  __asm__("sqrtsd %1, %0" : "=x"(x) : "x"(x));
+#else
+#error only aarch64 and x86 is supported
+#endif
+  return x;
+}
+
+static inline float sqrtf(float x) {
+#if defined(__aarch64__) || defined(_M_ARM64)
+  __asm__("fsqrt %s0, %s1" : "=w"(x) : "w"(x));
+#elif defined(__x86_64__) || defined(_M_X64)
+  __asm__("sqrtss %1, %0" : "=x"(x) : "x"(x));
+#else
+#error only aarch64 and x86 is supported
+#endif
+  return x;
+}
+
+uint16_t __truncdfhf2(double a) {
+  volatile float b = (float)a;
+  __fp16 c = (__fp16)b;
+  return *(uint16_t *)&c;
+}
+// from musl
```

Comment:
Why is this stuff needed?

Reply:
because llvm compiles stuff that isn't available as one instruction (like casting a double to half on most x86 platforms) into calls into compiler-rt that has a routines for it. For example entry point to compiler-rt's entry to impl of this - https://github.com/microsoft/compiler-rt/blob/master/lib/builtins/truncdfhf2.c . It is overcomplicated because double-half cast isn't really the same as double-float-half cast (double rounding afaik), so here instead of implementing idkhowcomplicatedpropperdoubletohalfis i just do it by converting double to float and then to half (return is uint16_t because there is no abi for __fp16 returns). Can't be made static inline like sqrt stuff because it looks like it gets inserted as a call only without respecting that the function with same symbol is available and marked as inline. The alternative is to parse ar files (because compiler-rt is a static library) and implement linking multiple .o together which i tried and it was complicated